Improves support for assigns clauses

feliperodri opened the pull request:
Assigns clauses are now able to handle structs and arrays.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs commented:
@NlightNFotis The code coverage is still showing the same spurious decrease as other PRs.  I think one of the changes to the build scripts has broken it.  Please can you have a look?

martin-cs commented:
@feliperodri Thank you for this work.  Would it be possible to break it into a number of smaller commits to make it easier to review?  As I understand it, this includes:

- A few changes to general supporting code
- A change to the parser to make assigns clauses attributes (I think)
- The corresponding updates to the regression tests plus some general cleaning and improvements
- Handling of structs and arrays in the assigns clauses
- Documentation updates

If these could be split up a bit, that would be great.

thomasspriggs commented:
This is not a full review. Just some initial observations. My apologies if I have jumped the gun by looking whilst it is still in draft status -

- CI is failing with `(core dumped)` errors. These will need to be fixed.
- A single large commit for an entire PR of this size is going to make it more difficult to review. It would be particularly useful if you could separate changes which are entirely stylistic, from those which make functional changes. That way reviewers can pay particular attention to the functional changes and the stylistic changes can be assessed on their own merit.
- Your changes include renaming existing variables to abbreviated forms. Our coding standards explicitly states that we should "Avoid abbreviations" for naming - https://github.com/diffblue/cbmc/blob/develop/CODING_STANDARD.md. Therefore such renaming should be removed from this PR, as it is against our standards and it has an adverse impact on readability. An additional issue with variable renaming, even where it is desirable, is that increases the size of the diff. This has the potential to cause additional merge resolution work, either for other people whose PRs are merged after yours, or for yourself if other people merge changes to the same areas before you. For this reason any renaming which you want to keep should be in a separate commit, which only includes the renaming and includes a commit message body justifying the renaming.
- `irep_idt` / `dstringt` includes an overload of `operator==`. This should be used in preference to converting both sides to C strings and using `std::strcmp`. Using the operator directly should be cleaner to read and more performant.
- I think I spotted `const` being removed from some of the declarations. I am not sure if you have added additional mutation, without looking more closely. But `const` should not be removed without good reason.

Good luck with finding the cause of the failing tests.

feliperodri commented:
@martin-cs @thomasspriggs I resolved all regression test issues and split up into a few commits, could you review it? Also, @thomasspriggs I revised the code for name abbreviations, misused of `std::strcmp`, and `const` removals.

tautschnig commented on regression/contracts/assigns_enforce_01/main.c:
```diff
@@ -1,5 +1,7 @@
 int foo(int *x) __CPROVER_assigns(*x)
-  __CPROVER_ensures(__CPROVER_return_value == *x + 5)
+  __CPROVER_ensures(__CPROVER_return_value == *x + 5);
```
For practical (bisecting) purposes: can you please make sure that the order of commits is such that at each commit all the tests (still) pass? Maybe that's the case, but I was expecting that some code changes would be necessary for this to work.

feliperodri replied:
Updates the git tree as suggested.

tautschnig commented on regression/contracts/assigns_enforce_02/main.c:
```diff
@@ -1,7 +1,9 @@
 
 int foo(int *x) __CPROVER_assigns(z)
-  __CPROVER_ensures(__CPROVER_return_value == *x + 5)
+  __CPROVER_ensures(__CPROVER_return_value == *x + 5);
```
Comment at random location: can you please add (meaningful) bodies to the commit messages? For this particular commit: "Moves all function contracts to function signatures" is a good subject, but then I'd like to be able to learn the answer to the "Why?" question in the body of the commit message.

feliperodri replied:
I updated messages in all commits for this PR. Does it look reasonable now?

tautschnig commented on regression/contracts/assigns_enforce_arrays_01/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+Checks ther assigns clause behavior when it reasons (indirectly) over a freshly-allocated variable.
```
Nit picking: "ther?"

feliperodri replied:
Fixed it.

tautschnig commented on regression/contracts/assigns_enforce_arrays_02/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that the instrumentation process does not duplicate function calls used as part of array index operations. That is, if a function call is used in the computation of the index of an array assignment, then intrumenting that statement with an assigns clause will not result in multiple function calls.
```
Nit pick 1: Please break lines at or before 80 characters.
Nit pick 2: s/intrumenting/instrumenting/

feliperodri replied:
Fixed it.

tautschnig commented on regression/contracts/assigns_enforce_arrays_12/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION FAILED$
+--
+--
+This test checks that verification fails whena function with an array assigns target calls a function which may assign below the allowable range.
```
s/whena/when a/

feliperodri replied:
Fixed it.

tautschnig commented on regression/contracts/assigns_enforce_arrays_13/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION FAILED$
+--
+--
+This test checks that verification fails whena function with an array assigns target calls a function which may assign above the allowable range.
```
s/whena/when a/

feliperodri replied:
Fixed it.

tautschnig commented on regression/contracts/assigns_enforce_arrays_14/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION FAILED$
+--
+--
+This test checks that verification fails whena function with an array assigns target calls a function which may assign both above or below the allowable range.
```
s/whena/when a/

feliperodri replied:
Fixed it.

tautschnig commented on regression/contracts/assigns_enforce_arrays_13/test.desc:
```diff
@@ -0,0 +1,9 @@
+KNOWNBUG
```
For this and other ones marked "KNOWNBUG": could you please clarify in the message at the botttom what is going wrong in that test at present?

tautschnig commented on doc/cprover-manual/assigns-clause.md:
```diff
@@ -1,52 +1,104 @@
+Here is described the syntax and semantics of the "assigns"
+clause as implemented in CBMC.  This construct allows the user to specify a list
+of memory locations which may be written within a function.  While an assigns
+clause may, in general, be interpreted with either "writes" or "modifies"
```
I'm not convinced this is an improvement over the previous documentation? The original language sounded better to me, but then English isn't my first language.

feliperodri replied:
Agreed. I reverted the changes in this paragraph.

tautschnig commented on doc/cprover-manual/assigns-clause.md:
```diff
@@ -1,52 +1,104 @@
+
+The assigns clause identifies a list of targets, which may be an identifier
+(e.g. `x`), a member of a struct (e.g. `myStruct.someMember` or
+`myStructPtr->otherMember`) , a dereferenced pointer (e.g. `*myPtr`), or an
```
Nit pick: remove space before comma

feliperodri replied:
Implemented as suggested.

tautschnig commented on doc/cprover-manual/assigns-clause.md:
```diff
@@ -1,52 +1,104 @@
-A special construct is introduced to specify assigns clauses. Its syntax is
-defined as follows.
+A special construct is introduced to specify assigns clauses.
+At the time of writing, its syntax is defined as follows.
```
I'd stick with the previous wording: it doesn't matter when this documentation was written, what matters to the reader is that it, at all times, correctly reflects the then-current implementation.

feliperodri replied:
Agreed. I reverted the changes in this paragraph.

tautschnig commented on src/util/pointer_offset_size.cpp:
```diff
@@ -492,6 +492,10 @@ optionalt<exprt> size_of_expr(const typet &type, const namespacet &ns)
+  else if(type.id() == ID_empty)
+  {
+    return from_integer(0, size_type());
+  }
```
This isn't correct (`void` has no defined size), but now I do wonder whether/why it even is necessary?

feliperodri replied:
It's wrong and unnecessary. Removed it.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,58 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
+  const exprt &op1() const
+  {
+    return operands()[1];
+  }
```
It should be possible to replace all this by
```
  // make op0 and op1 public
  using exprt::op0;
  using exprt::op1;

  const exprt &op2() const = delete;
  exprt &op2() = delete;
  const exprt &op3() const = delete;
  exprt &op3() = delete;
```

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,58 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
 }
 
+/// \brief A base class for int-range expressions
+class int_range_exprt : public expr_protectedt
```
Why does it matter that it's a range of integers? Couldn't we just have a `range_exprt` instead?

feliperodri replied:
I've changed to `range_exprt` as suggested.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,58 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
+  int_range_exprt(const irep_idt &_id, operandst _operands, typet _type)
+    : expr_protectedt(_id, std::move(_type))
+  {
+    operands() = std::move(_operands);
```
I don't think a constructor that requires exactly two entries in `_operands` should take an `operandst`. Instead, take two `exprt` parameters. Oh, and `_id` should be prescribed (e.g., `ID_range`).

feliperodri replied:
I updated the constructor as suggested.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,58 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
+/// \return Object of type \ref int_range_exprt
+inline const int_range_exprt &to_int_range_exprt(const exprt &expr)
+{
+  return static_cast<const int_range_exprt &>(expr);
```
Sanity checks need to happen here.

feliperodri replied:
I don't think there are any sanity checks for `range_exprt` at this point. Any suggestions?

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -341,7 +376,6 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
     }
 
     // do body
-
```
Unnecessary change

feliperodri replied:
Reverted it.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -379,6 +413,21 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
           throw 0;
         }
       }
+      else if(old_ct.parameters().size() > 0) // previous declaration
```
I'd use `!....empty()` instead of `.size() > 0`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -402,6 +451,29 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
+      }
+      if(old_ensures.is_not_nil())
+      {
+        new_symbol.type.move_to_named_sub(ID_C_spec_ensures, old_ensures);
```
I'd consistently use `move_to_named_sub`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -754,15 +825,15 @@ void c_typecheck_baset::typecheck_declaration(
+          contract.find(ID_C_spec_requires);
+      if(contract.find(ID_C_spec_ensures).is_not_nil())
+        new_symbol.type.add(ID_C_spec_ensures) =
+          contract.find(ID_C_spec_ensures);
```
Please use braces for multi-line `if` bodies.

feliperodri replied:
Fixed it.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -491,6 +491,40 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
+    exprt &upper = to_int_range_exprt(expr).op1();
+
+    if(lower.id() == ID_constant || lower.id() == ID_identifier)
+    {
```
What if they are not constant or an identifier?

feliperodri replied:
Currently, we only support constant, identifier and symbol. So, I added as a `PRECONDITION` rather than an if-statement.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -491,6 +491,40 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
+
+    if(lower.id() == ID_constant || lower.id() == ID_identifier)
+    {
+      dstringt c_type = lower.type().get(ID_C_c_type);
```
Don't use `dstringt` directly, use `const irep_idt &c_type = ...`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -491,6 +491,40 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
+    if(lower.id() == ID_constant || lower.id() == ID_identifier)
+    {
+      dstringt c_type = lower.type().get(ID_C_c_type);
+      if(c_type != ID_signed_int.c_str() && c_type != ID_unsigned_int.c_str())
```
There is no need for `.c_str()` (in fact: it defeats the whole purpose of `dstringt`)

feliperodri replied:
Fixed it.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -512,7 +512,9 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
-      if(upper_type != ID_signed_int.c_str() && upper_type != ID_unsigned_int.c_str())
+      if(
+        upper_type != ID_signed_int.c_str() &&
+        upper_type != ID_unsigned_int.c_str())
```
This change should be folded into the preceding commit (though will largely go away anyway once removing the `.c_str()`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -1032,17 +1081,39 @@ declaring_list:
+          PARSER.pop_scope();
+
+          // We are no longer in any function.
+          PARSER.set_function(PARSER.current_scope().prefix.substr(0, PARSER.current_scope().prefix.find_first_of(':')));
```
I believe you can safely use `find` instead of `find_first_of` (here and several times below).

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/parser_static.inc:
```diff
@@ -332,7 +332,8 @@ static void create_function_scope(const YYSTYPE d)
   ansi_c_declaratort &declarator=declaration.declarator();
 
-  irep_idt function_name=declarator.get_base_name();
+  std::string declarator_name(declarator.get_name().c_str());
```
Use `id2string(declarator.get_name())` instead of `.c_str()`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1471,7 +1471,6 @@ void dump_ct::cleanup_expr(exprt &expr)
 
     if(
       ns.follow(bu.type()).id() == ID_union &&
-      bu.source_location().get_function().empty() &&
```
This is a surprising change.

feliperodri replied:
This avoids a parsing error in the regression test `dump-union2`:
```c
name: main.c
Reading GOTO program from 'main.gb'
#include <assert.h>

// tag-S2
// file main.c line 3
struct S2;

// tag-U10
// file main.c line 9
union U10;

// tag-U4
// file main.c line 23
union U4;

// tag-U6
// file main.c line 17
union U6;


// __assert_rtn
// file /usr/include/assert.h line 76
_Noreturn void __assert_rtn(const char *, const char *, signed int, const char *);

struct S2
{
  // f0
  signed char f0;
  // f1
  unsigned short int f1;
};

union U10
{
  // f0
  unsigned short int f0;
  // f3
  struct S2 f3;
};

union U4
{
  // f0
  signed int f0 : 6;
  // f3
  signed int f3;
};

union U6
{
  // f0
  signed int f0 : 3;
};


// g_1197
// file main.c line 21
union U6 g_1197=byte_update_little_endian({  }, 0l, 1, signed int : 3);
// g_1256
// file main.c line 15
union U10 g_1256={ .f3={ .f0=0, .f1=0 }, .f0=(unsigned short int)-6l };
// g_1408
// file main.c line 28
union U4 g_1408=byte_update_little_endian({ .f3=0 }, 0l, -1l, signed long int);
// g_2110
// file main.c line 14
union U10 g_2110={ .f3={ .f0=0, .f1=0 }, .f0=53747 };

// main
// file main.c line 30
signed int main()
{
  /* assertion g_2110.f0 == 53747 */
  assert(!((signed long int)!((signed int)g_2110.f0 == 53747) != 0l));
  return 0;
}

main-mod.c: In function 'g_1197':
main-mod.c:57:1: error: syntax error before '{'
 union U6 g_1197=byte_update_little_endian({  }, 0l, 1, signed int : 3);
PARSING ERROR
```

The dumped C file would contain invalid `byte_update_little_endian` info.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -39,6 +39,7 @@ class return_value_visitort : public const_expr_visitort
 public:
   return_value_visitort() : const_expr_visitort()
   {
+    found = false;
```
Use `, found(false)` or (C11) declare the member with `bool found = false;`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -258,28 +276,14 @@ bool code_contractst::apply_contract(
+      function_symbol.location, function_id, function_symbol.mode);
 
     // Insert the non-deterministic assignment immediately before the call site.
+    int lines_to_iterate = assigns_havoc.instructions.size();
```
`std::size_t`, not `int`

feliperodri replied:
Fixed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -315,15 +319,28 @@ void code_contractst::code_contracts(
+    {
+      const code_function_callt &call = ins->get_function_call();
+
+      // TODO we don't handle function pointers
```
Replace the TODO by a `PRECONDITION`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -334,147 +351,157 @@ const symbolt &code_contractst::new_tmp_symbol(
   bool first_iter = true;
   exprt running = false_exprt();
-  for(auto aliasable : aliasable_references)
+  for(auto aliasble : aliasable_references)
```
Doesn't look like an improvement to me?

feliperodri replied:
Fixed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -968,3 +1002,731 @@ bool code_contractst::enforce_contracts(
+    }
+    else
+    {
+      result = and_exprt(result, current_target_compatible);
```
Use the `conjunction` helper function and just construct an `operandst` rather than building a nested tree of `and_exprt`.

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -968,3 +1002,731 @@ bool code_contractst::enforce_contracts(
   return fail;
 }
+
+/*********************************************
+ *  Assigns Clause Target - Scalar
+ ********************************************/
```
Not our way of formatting comments, and also not particularly informative (it's exactly the same as the name of the class).

feliperodri replied:
Removed it.

tautschnig reviewed (changes requested):
Comments for most of the commits (with some blockers). I am yet to properly review the largest of the commits, "Adds arrays and structs support for assigns clauses" - will do so after a first range of fixes has been applied.

martin-cs commented:
@feliperodri thanks for this and for separating the changes as requested.  I am happy to take another look but maybe it's best if I wait until you've had a chance to look at / action @tautschnig 's changes.

feliperodri commented:
> Comments for most of the commits (with some blockers). I am yet to properly review the largest of the commits, "Adds arrays and structs support for assigns clauses" - will do so after a first range of fixes has been applied.

@tautschnig thank you for all insightful comments. Could you take another review round?

feliperodri commented:
> @feliperodri thanks for this and for separating the changes as requested. I am happy to take another look but maybe it's best if I wait until you've had a chance to look at / action @tautschnig 's changes.

@martin-cs I agree. As soon as @tautschnig finish reviewing all changes, I can ping you for a final round.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -11,6 +11,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "c_typecheck_base.h"
 
+#include <iostream>
```
Probably a debugging left-over, please remove.

feliperodri replied:
Removed it.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -491,6 +492,41 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
+      error() << "Lower bound of range should be an integer, but got: "
+              << lower.type().pretty() << eom;
+      throw 0;
+    }
```
Does it really need to be `int`/`unsigned`? Wouldn't any signed/unsigned bitvector type work?

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -491,6 +492,41 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
+  {
+    // already type checked
+    exprt &lower = to_range_exprt(expr).op0();
+    exprt &upper = to_range_exprt(expr).op1();
```
I'd appreciate if the code handling `ID_range` could be moved to a function of its own, this function is way too long already.

feliperodri replied:
I created the method `typecheck_expr_range` to handle `ID_range`.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -491,6 +492,41 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
+      error() << "Upper bound of range should be an integer, but got: "
+              << upper.pretty() << eom;
+      throw 0;
+    }
```
As above, but most likely all you want is `implicit_typecast_arithmetic` (and have that do any error handling).

feliperodri replied:
I replaced both for `implicit_typecast_arithmetic`.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1470,9 +1470,7 @@ void dump_ct::cleanup_expr(exprt &expr)
-    if(
-      ns.follow(bu.type()).id() == ID_union &&
-      bu.source_location().get_function().empty())
+    if(ns.follow(bu.type()).id() == ID_union)
```
This perhaps is a symptom rather than a cause: if you've had to make this change to make the regression test pass, then the regression test suddenly has function names associated with global objects (that regression test doesn't have function-local unions).

feliperodri replied:
You're right, the problem is again with `create_function_scope($$);`. This part is also parsing unions
```cpp
        | type_specifier declarator
          post_declarator_attributes_opt
          {
            $2=merge($3, $2);
            
            // the symbol has to be visible during initialization
            init($$, ID_declaration);
            parser_stack($$).type().swap(parser_stack($1));
            PARSER.add_declarator(parser_stack($$), parser_stack($2));
            create_function_scope($$);
          }
          initializer_opt
          assigns_opt
          requires_opt
          ensures_opt
        {
          $$=$4;

          if(parser_stack($6).is_not_nil()) // Capture assigns clause
            parser_stack($$).add(ID_C_spec_assigns).swap(parser_stack($6));

          // Capture code contract
          if(parser_stack($7).is_not_nil())
            parser_stack($$).add(ID_C_spec_requires).swap(parser_stack($7));
          if(parser_stack($8).is_not_nil())
            parser_stack($$).add(ID_C_spec_ensures).swap(parser_stack($8));

          // add the initializer
          if(parser_stack($5).is_not_nil())
          {
            to_ansi_c_declaration(parser_stack($$)).add_initializer(parser_stack($5));
          }

          // Kill the scope that 'function_head' creates.
          PARSER.pop_scope();

          // We are no longer in any function.
          PARSER.set_function(PARSER.current_scope().prefix.substr(0, PARSER.current_scope().prefix.find(':')));
        }
```
So when we `set_function` in `create_function_scope`, we also endup setting for unions. Any suggestion on how to handle this case?

tautschnig replied:
Let's discuss how we can refactor this PR. These changes to the C parser are pretty scary, but I don't want to block the entire set of improvements because of this part.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,49 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
+/// \return Object of type \ref range_exprt
+inline const range_exprt &to_range_exprt(const exprt &expr)
+{
+  return static_cast<const range_exprt &>(expr);
```
Prefix this by `PRECONDITION(expr.id() == ID_range);`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,49 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
+/// \return Object of type \ref range_exprt
+inline range_exprt &to_range_exprt(exprt &expr)
+{
+  return static_cast<range_exprt &>(expr);
```
Prefix this by `PRECONDITION(expr.id() == ID_range);`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -156,14 +172,15 @@ bool code_contractst::has_contract(const irep_idt fun_name)
-  // TODO: handle function pointers.
+  PRECONDITION(call.function().id() != ID_dereference);
   if(call.function().id() != ID_symbol)
     return false;
```
I believe the new `PRECONDITON` should make this `return false;` unreachable. Perhaps turn all this into `PRECONDITION(call.function().id() == ID_symbol);`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -223,16 +240,16 @@ bool code_contractst::apply_function_contract(
     {
       symbol_exprt p(p_it->get_identifier(), p_it->type());
-      replace.insert(p, *a_it);
+      replace.insert(p, *arguments_it);
```
As much as I agree this is an improvement: the commit is huge already, and would benefit from not having extra changes in it.

feliperodri replied:
I reverted this change. We can have another PR with clean ups.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -11,10 +11,13 @@ Date: February 2016
 /// Verify and use annotated invariants and pre/post-conditions
 
+#include "code_contracts.h"
+#include "loop_utils.h"
```
Nit pick: we usually have the header file for the particular implementation at the top, but really only this one. Other headers from the same directory go at the very end of the list of includes. Net result: `#include "loop_utils.h"` should stay where it was.

feliperodri replied:
Moved back to where it was.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -306,13 +309,33 @@ void code_contractst::apply_loop_contract(
+      }
+    }
+  }
+#endif
```
You're introducing commented-out code: if you have good reasons to do so, please spell them out in a comment.

feliperodri replied:
I removed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -306,13 +309,33 @@ void code_contractst::apply_loop_contract(
   const source_locationt &source_location,
   const irep_idt &function_id,
-  const irep_idt &mode)
+  const irep_idt &mode) const
```
How is that possible when `get_fresh_aux_symbol` modifies `symbol_table`?

feliperodri replied:
I removed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
   for(auto aliasable : aliasable_references)
   {
-    exprt left_ptr = address_of_exprt{lhs};
+    exprt left_ptr = exprt(ID_address_of, pointer_type(lhs.type()), {lhs});
```
Is this an improvement? Also, this expression is the same in each loop iteration, it should probably just be created once.

feliperodri replied:
Removed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
   std::vector<exprt> &aliasable_references)
 {
   bool first_iter = true;
   exprt running = false_exprt();
```
This function could be made more succinct by using `exprt::operandst operands; operands.reserve(aliasable_references.size();` here and each loop iteration doing `operands.push_back(compatible);` with a final `return disjunction(operands);`

feliperodri replied:
I completely re-wrote this function to remove dead code. This also fixes https://github.com/diffblue/cbmc/issues/5528.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
+    exprt same_offset = equal_exprt(lhs, typecast_exprt(right_ptr, lhs.type()));
+
+    auto left_size = size_of_expr(pointer_offset(left_ptr).type(), ns);
+    const exprt &compatible = same_offset;
```
`left_size` is never used, so I'm wondering whether this is doing what was intended?

feliperodri replied:
Removed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
-    instruction_iterator->is_assign(),
-    "The first argument of instrument_assigns_statement should always be"
+    instruction_it->is_assign(),
+    "The first argument of instrument_assign_statement should always be"
```
It's not actually the first argument anymore.

feliperodri replied:
Updated to refer to the first instruction.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
   }
-  exprt alias_expr = create_alias_expression(assigns, lhs, assigns_references);
+
+  exprt left_ptr = exprt(ID_address_of, pointer_type(lhs.type()), {lhs});
```
`address_of_exprt`?

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
+  exprt cast_alias = not_exprt(binary_predicate_exprt(
+    typecast_exprt(
+      typecast_exprt(not_exprt(alias_expr), signed_long_int_type()),
+      signed_long_int_type()),
```
The outer `typecast_exprt` seems redundant?

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
+      typecast_exprt(not_exprt(alias_expr), signed_long_int_type()),
+      signed_long_int_type()),
+    ID_notequal,
+    constant_exprt(irep_idt(dstringt("0")), signed_long_int_type())));
```
`from_integer(0, signed_long_int_type())`

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
+  }
+
+  exprt alias_expr = assigns_clause.alias_expression(lhs);
+  exprt cast_alias = not_exprt(binary_predicate_exprt(
```
I'm not really sure what this expression is supposed to tell. Why is a Boolean (`alias_expr`) being converted to integer types here?

feliperodri replied:
The entire cast is useless. I removed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
 
 void code_contractst::instrument_call_statement(
-  goto_programt::instructionst::iterator &instruction_iterator,
+  goto_programt::instructionst::iterator &instruction_it,
```
Not obviously an improvement?

feliperodri replied:
Reverted it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
-  std::set<exprt> &freely_assignable_exprs)
+  const irep_idt &function_id,
+  std::set<dstringt> &freely_assignable_symbols,
+  assigns_clauset &assigns_clause)
```
I'd recommend a consistent order of parameters across the various `instrument_...` functions.

feliperodri replied:
I created a GitHub issue (https://github.com/diffblue/cbmc/issues/5941) to tackle this on a separate PR. I can probably remove `function_id` from the function parameters for all functions, but I think it's better do this as part of a general clean up in this module.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
+      else
+      {
+        log.error() << "Malloc is called but the result is not cast. "
+                    << "Excluding result from the assignable memory frame ."
```
Nit pick: space before "."

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
 
-  // TODO we don't handle function pointers
+  PRECONDITION(call.function().id() != ID_dereference);
   if(call.function().id() == ID_symbol)
```
As above: use `PRECONDITION(call.function().id() == ID_symbol);`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -323,147 +346,157 @@ const symbolt &code_contractst::new_tmp_symbol(
+    // to a result variable. We iterate one line forward to grab the result of
+    // the malloc once it is cast.
+    local_instruction_it++;
+    if(local_instruction_it->is_assign())
```
How does this relate to what the commit message talks about? I'm sure these are useful changes, but it would be so much easier to understand and review if these changes were paired up with a regression test that is made to work by these changes.

feliperodri replied:
This change is necessary to handle the following cases:
 - `assigns_enforce_malloc_01`
 - `assigns_enforce_multi_file_02`
 - `assigns_enforce_scoping_02`
 - `assigns_enforce_structs_01`
 - `assigns_enforce_structs_02`
 - `assigns_enforce_structs_03`

But I understand your point and I'll keep in mind to future contributions. Each commit should contain a specific change and a set of regression tests (if applicable) to map the new behavior.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -620,72 +634,93 @@ bool code_contractst::add_pointer_checks(const std::string &function_name)
 
   // Create a list of variables that are okay to assign.
-  std::set<exprt> freely_assignable_exprs;
+  std::set<dstringt> freely_assignable_symbols;
```
Use `irep_idt`, not `dstringt`

feliperodri replied:
Implemented as suggested.

tautschnig commented on scripts/expected_doxygen_warnings.txt:
```diff
@@ -1,16 +1,16 @@
-/home/runner/work/cbmc/cbmc/src/solvers/sat/satcheck_glucose.cpp:246: warning: no matching class member found for 
+/home/runner/work/cbmc/cbmc/src/solvers/sat/satcheck_glucose.cpp:253: warning: no matching class member found for 
```
Surprising change as this PR doesn't obviously change this file.

feliperodri replied:
Do you think changes in other files could impact this one?

tautschnig reviewed (changes requested):
Thank you for all the updates. I got a bit further into the largest commit, but there seem to be a lot of not-obviously-related changes baked into it.

kroening commented:
The changes to the grammar are indeed very scary, and should likely be pulled out of this PR.

The right place to do this appears to be `postfixing_abstract_declarator`; I'll try.

kroening commented:
Look at https://github.com/diffblue/cbmc/pull/5946 for a proposal how to change the grammar.

feliperodri committed:
Adds arrays and structs support for assigns clause

Assigns clauses should handle multiple targets.
Previously, it only supported identifiers (e.g., variable `x`)
and dereferenced pointers (e.g., *myPtr). So, we expand its support to
also handle members of a struct (e.g. `myStruct.someMember` or
`myStructPtr->otherMember`) and array ranges (e.g. `myArr[5, lastIdx]`).

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
Updates assigns clause description in CPROVER manual

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
Adds regression test to cover code contracts support

Code contracts now supports assigns clause with structs and array ranges.
It also supports contracts specified across multiple source files.
So, we added regression tests to cover these capabilites.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri commented:
> The changes to the grammar are indeed very scary, and should likely be pulled out of this PR.
> 
> The right place to do this appears to be `postfixing_abstract_declarator`; I'll try.

@tautschnig @kroening I completely removed the changes in this PR related to updating the grammar to support contracts on function declarations. I'll take a look at Daniel's proposal.

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +533,63 @@ target_list:
         ;
 
 target:
+         deref_target
```
Nit pick: indent by one more space (so as to align with `target` in the next line).

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +533,63 @@ target_list:
+        ;
+
+deref_target:
+         member_target
```
Nit pick: indent by one more space (so as to align with `'*'` in the next line).

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +533,63 @@ target_list:
         ;
 
+member_target:
+         primary_target
```
Nit pick: indent by one more space (so as to align with `member_target` in the next line).

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +533,63 @@ target_list:
+        ;
+
+primary_target:
+         identifier
```
Nit pick: indent by one more space (so as to align with `'('` in the next line).

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +533,63 @@ target_list:
+        { $$ = $2; }
+        ;
+
+
```
Nit pick: just a single blank line will do

feliperodri replied:
Removed it.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -11,10 +11,12 @@ Date: February 2016
-#include <unordered_set>
 
 #include <analyses/local_may_alias.h>
+#include <ansi-c/expr2c.h>
```
Please remove this and instead include `util/format_type.h` and replace all calls to `type2c(...)` by `format(...)`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -219,7 +235,7 @@ bool code_contractst::apply_function_contract(
 
   // Replace formal parameters
-  code_function_callt::argumentst::const_iterator a_it=
+  code_function_callt::argumentst::const_iterator a_it =
```
Nit pick: now that you reverted the change, please also revert the formatting improvement.

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -79,6 +83,14 @@ class code_contractst
+    const irep_idt &function_id,
+    const irep_idt &mode);
+
+  const namespacet &get_namespace() const;
```
Just to keep things more consistent across the code base: would it be possible to get rid of this method and instead pass along the `ns` member wherever `get_namespace` currently has to be used?

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -100,56 +112,59 @@ class code_contractst
     exprt &assigns,
-    std::vector<exprt> &original_references,
-    std::set<exprt> &freely_assignable_exprs);
+    std::set<dstringt> &freely_assignable_symbols,
```
Use `irep_idt` instead of `dstringt` (you might want to search for `dstringt` and make sure it doesn't appear anywhere in your code).

feliperodri replied:
I searched for all instance of `dstringt` in `code_contracts.*` files an replaced it all with `irep_idt`.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -100,56 +112,59 @@ class code_contractst
 
   /// Inserts an assertion statement into program before the function call at
-  /// ins_it, to ensure that any memory which may be written by the call is
+  /// instruction_it, to ensure that any memory which may be written by the call is
```
Nit pick: probably an improvement, but also adds noise to this already-very-large PR.

feliperodri replied:
Reverted it.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -100,56 +112,59 @@ class code_contractst
-    const irep_idt &func_id,
-    goto_programt &created_decls,
+    const irep_idt &function_id,
+    std::set<dstringt> &freely_assignable_symbols,
```
See above: make sure `dstringt` doesn't occur.

feliperodri replied:
Replaced it with `irep_idt`.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -170,4 +185,175 @@ class code_contractst
+class assigns_clause_targett
+{
+public:
+  enum TargetType
```
Don't use CaMeLcAsE, not even for enums.

feliperodri replied:
ReMoVeD iT.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -170,4 +185,175 @@ class code_contractst
+    TargetType type,
+    const exprt object_ptr,
+    const code_contractst &contract,
+    messaget &log_parameter)
```
This is one of those that may need to require a `const namespacet &` as a further parameter to help get rid of `get_namespace()`.

feliperodri replied:
I created this GitHub issue https://github.com/diffblue/cbmc/issues/5950 to remove all instances of `get_namespace()` from the code contracts module. This should be part of the refactor.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -170,4 +185,175 @@ class code_contractst
+  static const exprt pointer_for(const exprt &exp)
+  {
+    const exprt &left_ptr =
+      exprt(ID_address_of, pointer_type(exp.type()), {exp});
```
Use `address_of_exprt` and move this code below the `if` to avoid unnecessary object construction.

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -170,4 +185,175 @@ class code_contractst
+    return init_block;
+  }
+
+  static const exprt pointer_for(const exprt &exp)
```
No need for `const` here.

feliperodri replied:
Removed it.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -170,4 +185,175 @@ class code_contractst
+  const TargetType target_type;
+
+protected:
+  const exprt pointer_object;
+  const code_contractst contract;
```
Should these perhaps be `const &`?

feliperodri replied:
`contract` is but `pointer_object` no.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -170,4 +185,175 @@ class code_contractst
+  goto_programt havoc_code(source_locationt location) const;
+
+protected:
+  int lower_bound;
+  int upper_bound;
```
Both of these should be `mp_integer`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +950,723 @@ bool code_contractst::enforce_contracts(
+    }
+    if(
+      left_size.value().get(ID_value) ==
+      right_size.value().get(ID_value).c_str())
```
Remove `.c_str()`; in fact you'll really want `*left_size == *right_size`.

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +950,723 @@ bool code_contractst::enforce_contracts(
+  bool first_it = true;
+  for(symbol_exprt symbol : local_standin_variables)
+  {
+    typet ptr_concrete_type = dereference_exprt(ptr).type();
```
`const typet &ptr_concrete_type = to_pointer_type(ptr.type()).subtype();`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +950,723 @@ bool code_contractst::enforce_contracts(
+
+exprt assigns_clause_struct_targett::alias_expression(const exprt &ptr)
+{
+  exprt running = false_exprt();
+  bool first_it = true;
```
Replace these by `exprt::operandst disjuncts; disjuncts.reserve(local_standin_variables.size());` and ... (continues below)

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +950,723 @@ bool code_contractst::enforce_contracts(
+  {
+    typet ptr_concrete_type = dereference_exprt(ptr).type();
+    auto left_size = size_of_expr(ptr_concrete_type, contract.get_namespace());
+    typet standin_concrete_type = dereference_exprt(symbol).type();
```
`const typet &standin_concrete_type = to_pointer_type(symbol.type()).subtype();`

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +950,723 @@ bool code_contractst::enforce_contracts(
+      exprt same_offset =
+        equal_exprt(pointer_offset(ptr), pointer_offset(symbol));
+
+      const exprt &compatible = and_exprt(same_obj, same_offset);
+      if(first_it)
+      {
+        running = compatible;
+        first_it = false;
+      }
+      else
+      {
+        running = or_exprt(running, compatible);
+      }
```
... (continues from above): replace this by `disjuncts.push_back(and_exprt{same_obj, same_offset});` ... (continues below)

feliperodri replied:
Implemented as suggested.

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +950,723 @@ bool code_contractst::enforce_contracts(
+    }
+  }
+
+  return running;
```
`return disjunction(disjuncts);`

feliperodri replied:
Implemented as suggested.

tautschnig reviewed (approved):
I have not read all of the changes to `code_contracts.cpp`, but have provided some suggestions for improvement. Most likely that bit of code needs a complete overhaul anyway. All other parts are ok except for nit picks.

feliperodri commented:
@martin-cs and @thomasspriggs can I get one more review on this PR?

NlightNFotis commented on src/ansi-c/scanner.l:
```diff
@@ -1517,6 +1517,7 @@ __decltype          { if(PARSER.cpp98 &&
 "&&"            { loc(); return TOK_ANDAND; }
 "||"            { loc(); return TOK_OROR; }
 "..."           { loc(); return TOK_ELLIPSIS; }
+".."            { loc(); return TOK_RANGE; }
```
What's the purpose of this token? It doesn't seem to be used in the Parser, or am I mistaken?

feliperodri replied:
I removed it.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -60,6 +63,22 @@ class return_value_visitort : public const_expr_visitort
+    log.error().source_location = type.source_location();
+    log.error() << "unable to determine size of type: " << format(type)
+                << messaget::eom;
+    throw 0;
```
Hm, personally I'm not a huge fan of this idiom in our source code.

I think this is better served as an invariant, in particular a `CHECK_RETURN`. What do you think?

feliperodri replied:
I don't have a strong opinion about this change, so I implemented as suggested.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -152,16 +171,15 @@ bool code_contractst::has_contract(const irep_idt fun_name)
-  // TODO: handle function pointers.
-  if(call.function().id() != ID_symbol)
-    return false;
+  PRECONDITION(call.function().id() == ID_symbol);
```
Is `function_id == call.function().id()`? If yes, does it make sense for us to check this too?

feliperodri replied:
Not necessary, the `function_id` may change if we want to replace the contract instead of enforcing it.

NlightNFotis commented on doc/cprover-manual/assigns-clause.md:
```diff
@@ -1,19 +1,17 @@
+as their support would require changes to CBMC's memory model.
 
 
 ##### Syntax
```
Small nitpick: Can I ask that headings start with H1 (`#`), and then descend one level each?

It would also be great if we can get an empty line under each heading.

Good documentation by the way 👍🏻

feliperodri replied:
Implemented as suggested.

NlightNFotis commented on doc/cprover-manual/assigns-clause.md:
````diff
@@ -24,29 +22,82 @@ defined as follows.
+`foo`.   
+
+
+```
````
Nitpick: By the way, in the beginning of a code fence you can declare the language, e.g `c`, so that syntax highlighting works in markdown editors and viewers.

feliperodri replied:
Good catch! Implemented as suggested.

NlightNFotis commented on regression/contracts/assigns_enforce_arrays_06/main.c:
```diff
@@ -0,0 +1,12 @@
+void assign_out_under(int a[], int len) __CPROVER_assigns(a[8])
+{
+  a[7] = 5;
```
Is this effectively the same test as `assigns_enforce_arrays_05` or do I think so?

feliperodri replied:
It's quite similar. In this case we cant to see if we report failure correctly when the specification doesn't have a range, but a single cell (i.e., `a[8]` instead of `a[2, 5]`).

NlightNFotis commented on regression/contracts/assigns_enforce_arrays_06/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+--
+Checks whether verification fails when an array is assigned at an index
+above its allowed index.
```
Is the wording here wrong? It seems it's below the allowed index.

feliperodri replied:
Good catch! You're correct, this is probably a c&p error.

NlightNFotis commented on regression/contracts/assigns_enforce_arrays_08/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+--
+Checks whether verification fails when an array is assigned at an index
+below its allowed index.
```
Ditto.

feliperodri replied:
Fixed it.

NlightNFotis commented on regression/contracts/assigns_enforce_arrays_12/test.desc:
```diff
@@ -0,0 +1,10 @@
+^VERIFICATION FAILED$
+--
+--
+Checks whether verification fails when a function with an array assigns target
```
Is there any warning printed for this case?

If yes, does it make sense to pattern match for it?

feliperodri replied:
We currently do not have any warnings related to this issue. Should I create a GitHub issue to add this in future PRs?

NlightNFotis reviewed (commented):
Reasonable so far.

I only have some questions and some minor requests.

NlightNFotis commented on src/ansi-c/parser.y:
```diff
@@ -101,6 +101,7 @@ extern char *yyansi_ctext;
 %token TOK_NE        "!="
 %token TOK_ANDAND    "&&"
 %token TOK_OROR      "||"
+%token TOK_RANGE     ".."
```
I'm sorry to be a pain, but can this also be removed?

feliperodri replied:
Not a pain at all, thank you for the thoroughly review! Removed it.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -454,41 +427,38 @@ void code_contractst::instrument_call_statement(
+  auto called_func = goto_functions.function_map.find(called_name);
+  if(called_func == goto_functions.function_map.end())
+  {
+    log.error() << "Could not find function '" << called_name
```
Sorry, can this also be changed to an invariant?

feliperodri replied:
Unnecessary condition. Removed it.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -454,41 +427,38 @@ void code_contractst::instrument_call_statement(
+  if(called_assigns.is_nil()) // Called function has no assigns clause
+  {
+    // Fail if called function has no assigns clause.
+    log.error() << "No assigns specification found for function '"
```
Ditto.

feliperodri replied:
We don't need this error log here, since this is a valid path. Updated it.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +940,713 @@ bool code_contractst::enforce_contracts(
+    if(!left_size.has_value())
+    {
+      log.error().source_location = ptr.source_location();
+      log.error() << "unable to determine size of type (lhs): "
```
Ditto.

feliperodri replied:
Implemented as suggested.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +940,713 @@ bool code_contractst::enforce_contracts(
+    }
+    if(!right_size.has_value())
+    {
+      log.error().source_location = ptr.source_location();
```
Ditto.

feliperodri replied:
Implemented as suggested.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -961,3 +940,713 @@ bool code_contractst::enforce_contracts(
+    ptr_offset);
+  exprt in_range = and_exprt(in_range_lower, in_range_upper);
+
+  // Troublesome
```
What is troublesome here? And why? Does it need to be guarded by an Invariant maybe?

feliperodri replied:
Fixed it.

NlightNFotis commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -319,133 +316,109 @@ const symbolt &code_contractst::new_tmp_symbol(
+      }
+      else
+      {
+        log.error() << "Malloc is called but the result is not cast. "
```
Does this also make sense for us to eliminate the if/else altogether, and just have an invariant guarding `rhs.id()`?

feliperodri replied:
Yes, implemented as suggested.

NlightNFotis reviewed (commented):
Sorry, some small comments and otherwise it's good to go for me.

feliperodri commented:
@NlightNFotis could you take another look?

NlightNFotis reviewed (approved):
Looks good to me, thanks for the implementation of the comment suggestions.

hannes-steffenhagen-diffblue commented on src/util/std_expr.h:
```diff
@@ -824,6 +824,51 @@ inline multi_ary_exprt &to_multi_ary_expr(exprt &expr)
+  }
+
+  // make op0 and op1 public
+  using exprt::op0;
```
*strong* preference to give these sensible names. You already call them "lower" and "upper", what's wrong with just using these as accessors instead of the opaque "op0" and "op1"?

feliperodri replied:
Implemented as suggested.

hannes-steffenhagen-diffblue commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +532,62 @@ target_list:
+        | integer
+        ;
+
+array_range:
```
Not a blocker, but a comment from me: I'm quite worried about the proliferation of nonstandard extensions. We're getting into a situation where we have to support gcc extensions, clang extensions, visual studio extensions on top of our own modifications to the language...

This could've been done without any parser changes at the relatively small cost of a less cute syntax, e.g. `__CPROVER_array_range(array, lower, upper)` or something like that. Looks bad, but less problematic syntactically.

feliperodri replied:
For now, I'll keep using this annotation, but that's a valid concern. Thus, I created the issue https://github.com/diffblue/cbmc/issues/5966 to discuss whether we should continue to use this annotation for array ranges or only adopt annotations that do not require changes to the parser.

feliperodri replied:
Based on the discussion on #5966 I reverted all the parser changes (to support array ranges) from this PR. We should have a syntax similar to ACSL, but that should be added on a separate PR. I kept the code on the assigns clause size for this future support.

hannes-steffenhagen-diffblue commented on src/ansi-c/parser.y:
```diff
@@ -532,15 +532,62 @@ target_list:
+        ;
+
+array_range:
+        array_index ',' array_index
```
Potential 🙅 

I'm reviewing commit-by-commit so maybe this is explained later on, but is the syntax we're adding here `array[lower_bound, upper_bound]`? In which case this needs to be changed, that is already totally valid syntax in C ([comma operator](https://en.wikipedia.org/wiki/Comma_operator)).

Maybe this is explained somewhere else in the conversation above, if so I'm sorry if I missed it!

NlightNFotis replied:
Personally, it's my position that this is acceptable.

 Unless I'm mistaken, from the grammar, this particular production derives only from a particular context (`assigns_opt` which part of the contract productions during a function definition), so even if an argument can be made about the overloading of the operator itself being unwanted, it's within a very narrow context, in order to allow for syntax that is congruent to the mathematical notation.

hannes-steffenhagen-diffblue replied:
@NlightNFotis considering it'd be a pretty trivial change to make this not be a problem anymore (either use a custom operator like `..` or just accept slightly uglier syntax that doesn't need any parser changes at all like I suggest above) I can't really agree to that.

feliperodri replied:
I've updated our annotation to adopt `[lower .. upper]` instead of the comma operator.

hannes-steffenhagen-diffblue reviewed (changes requested):
Unless I'm mistaken the changes here lead to a conflict in the parser w.r.t. to the comma operator. If I'm mistaken please explain.

feliperodri committed:
Updates list of expected Doxygen warnings

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

hannes-steffenhagen-diffblue reviewed (approved):
Removing my block because the specific issue I was complaining about was addressed. Larger point about syntax extensions remains, but doesn't need to be addressed at this very moment.
